README
NAME
    Setup::Test - Various simple setup routines for testing

VERSION
    version 0.10

SYNOPSIS
     use Setup::Unix::Group 'setup_text_case';

     # simple usage (doesn't save undo data)
     my $text = 'foo bar baz';
     my $res = setup_text_case text_ref => \$text, case => 'upper';
     die unless $res->[0] == 200 || $res->[0] == 304;

     # perform setup and save undo data (undo data should be serializable)
     $res = setup_text_case ..., -undo_action => 'do';
     die unless $res->[0] == 200 || $res->[0] == 304;
     my $undo_data = $res->[3]{undo_data};

     # perform undo
     $res = setup_text_case ..., -undo_action => "undo", -undo_data=>$undo_data;
     die unless $res->[0] == 200 || $res->[0] == 304;

DESCRIPTION
    This module provides simple setup functions, useful for testing
    purposes.

    This module uses Log::Any logging framework.

    This module has Rinci metadata.

SEE ALSO
    Setup

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Setup-Test

0.10    2012-07-04 (SHARYANTO)

        No functional changes. Adjust to Perinci::Sub::Gen::Undoable 0.09.


0.09    2012-06-23 (SHARYANTO)

        No functional changes. Adjust to Perinci::Sub::Gen::Undoable 0.08.


0.08    2012-06-22 (SHARYANTO)

        No functional changes. Now uses Perinci::Sub::Gen::Undoable.


0.07    2012-03-28 (SHARYANTO)

        No functional changes. Use newer Test::Setup.


0.06    2012-03-28 (SHARYANTO)

        No functional changes. Use Test::Setup.


0.05    2012-03-06 (SHARYANTO)

        No functional changes. Add missing dependency to Log::Any [CT].


0.04    2012-03-02 (SHARYANTO)

        No functional changes. Replace Sub::Spec with Rinci.


0.03    2011-05-18 (SHARYANTO)

        No functional changes. Changed logging levels.


0.02    2011-05-17 (SHARYANTO)

        [BUG FIXES]

        - Fixed typo in code.


0.01    2011-05-17 (SHARYANTO)

        - First release.


t/testlib.pl
use 5.010;
use strict;
use warnings;

#use File::chdir;
#use File::Slurp;
#use File::Temp qw(tempdir);
use Setup::Test qw(setup_text_case);
use Test::More 0.96;
use Test::Setup qw(test_setup);

sub setup {
    #$::tmp_dir = tempdir(CLEANUP => 1);
    #$CWD = $::tmp_dir;

    #diag "tmp dir = $::tmp_dir";
}

sub teardown {
    done_testing();
    if (Test::More->builder->is_passing) {
        #diag "all tests successful, deleting temp files";
        #$CWD = "/";
    } else {
        #diag "there are failing tests, not deleting temp files";
    }
}

sub test_setup_text_case {
    my %tstargs = @_;

    my %tsargs = ();

    for (qw/name dry_do_error do_error set_state1 set_state2 prepare cleanup/) {
        $tsargs{$_} = $tstargs{$_};
    }
    $tsargs{function} = \&setup_text_case;


    my %fargs = (%{$tstargs{args} // {}},
             );
    $tsargs{args} = \%fargs;

    my $check = sub {
        my %cargs = @_;

        if (defined $cargs{text}) {
            is(${$fargs{text_ref}}, $cargs{text}, "text")
                or diag explain(${$fargs{text_ref}});
        }
    };

    $tsargs{check_setup}   = sub { $check->(%{$tstargs{check_setup}}) };
    $tsargs{check_unsetup} = sub { $check->(%{$tstargs{check_unsetup}}) };
    if ($tstargs{check_state1}) {
        $tsargs{check_state1} = sub { $check->(%{$tstargs{check_state1}}) };
    }
    if ($tstargs{check_state2}) {
        $tsargs{check_state2} = sub { $check->(%{$tstargs{check_state2}}) };
    }

    test_setup(%tsargs);
}

1;


lib/Setup/Test.pm
package Setup::Test;

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use Perinci::Sub::Gen::Undoable 0.09 qw(gen_undoable_func);

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(setup_text_case);

our $VERSION = '0.10'; # VERSION

our %SPEC;

my $res = gen_undoable_func(
    name     => __PACKAGE__ . '::setup_text_case',
    summary  => "Change text case",
    description => <<'_',

On do, will change text case to UPPER/lower/Title Case. Will save the original
text for undo.

On undo, will restore the original text.

This function does not support transaction because it uses scalar references
which is not serializable to JSON (format used by transaction manager).

_
    tx   => {use=>0},
    args => {
        text_ref => {
            schema  => 'str*', # XXX actually ref to str, not str
            summary => 'Reference to text',
        },
        case => {
            schema  => ['str*' => {in=>[qw/upper lower title/]}],
            summary => 'Case style',
        },
    },

    check_args => sub {
        my $args = shift;
        defined($args->{text_ref}) or return [400, "Please specify text_ref"];
        ref($args->{text_ref}) eq 'SCALAR'
            or return [400, "Invalid text_ref: must be ref to a scalar"];
        $args->{case} or return [400, "Please specify case"];
        $args->{case} =~ /\A(upper|lower|title)\z/
            or return [400, "Invalid case: must be upper/lower/title"];
        [200, "OK"];
    },

    build_steps => sub {
        my $args = shift;
        [200, "OK", [[case => $args->{case}]]];
    },

    steps => {
        case => {
            summary => 'Change text case',
            check_or_fix => sub {
                my ($which, $args, $step, $undo) = @_;
                my $tr   = $args->{text_ref};
                my $case = $args->{case};
                my $correct;
                if ($case eq 'upper') {
                    $correct = uc($$tr);
                } elsif ($case eq 'lower') {
                    $correct = lc($$tr);
                } elsif ($case eq 'title') {
                    ($correct = $$tr) =~ s/\b(\w)(\w*)\b/uc($1).lc($2)/eg;
                }
                if ($which eq 'check') {
                    return [200, "OK"] if $$tr eq $correct;
                    return [200, "OK", ["set", $$tr]];
                } else { # fix
                    $$tr = $correct;
                    [200, "OK"];
                }
            },
        },

        set => {
            summary => 'Set (restore) text value',
            check => sub {
                my ($args, $step) = @_;
                [200, "OK", ["case", $args->{case}]];
            },
            fix => sub {
                my ($args, $step, $undo) = @_;
                my $tr = $args->{text_ref};
                $$tr = $step->[1];
                [200, "OK"];
            },
        },
    }
);

die "Can't generate function: $res->[0] - $res->[1]" unless $res->[0] == 200;
$SPEC{setup_text_case} = $res->[2]{meta};

1;
# ABSTRACT: Various simple setup routines for testing


__END__
=pod

=head1 NAME

Setup::Test - Various simple setup routines for testing

=head1 VERSION

version 0.10

=head1 SYNOPSIS

 use Setup::Unix::Group 'setup_text_case';

 # simple usage (doesn't save undo data)
 my $text = 'foo bar baz';
 my $res = setup_text_case text_ref => \$text, case => 'upper';
 die unless $res->[0] == 200 || $res->[0] == 304;

 # perform setup and save undo data (undo data should be serializable)
 $res = setup_text_case ..., -undo_action => 'do';
 die unless $res->[0] == 200 || $res->[0] == 304;
 my $undo_data = $res->[3]{undo_data};

 # perform undo
 $res = setup_text_case ..., -undo_action => "undo", -undo_data=>$undo_data;
 die unless $res->[0] == 200 || $res->[0] == 304;

=head1 DESCRIPTION

This module provides simple setup functions, useful for testing purposes.

This module uses L<Log::Any> logging framework.

This module has L<Rinci> metadata.

=head1 SEE ALSO

L<Setup>

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => "0.96",
    "Test::Setup" => "1.01"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Various simple setup routines for testing",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Setup-Test",
  "dist_version" => "0.10",
  "license" => "perl",
  "module_name" => "Setup::Test",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Log::Any" => 0,
    "Perinci::Sub::Gen::Undoable" => "0.09",
    "Rinci" => "1.1.0",
    "Setup" => "1.0.0",
    "perl" => "5.010000"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
